experimental/LieAlgebras/src/WeylGroup.jl

```diff
@@ -290,3 +290,130 @@ function parabolic_subgroup_with_projection(
+
+julia> irreducible_factors(W)
+2-element Vector{WeylGroup}:
+ Weyl group of root system of type A3
+ Weyl group of root system of type B4
+
+julia> U, emb, proj = irreducible_factors(W; morphisms=true)
+(WeylGroup[Weyl group of root system of type A3, Weyl group of root system of type B4], Map{WeylGroup, WeylGroup}[Map: Weyl group -> W, Map: Weyl group -> W], Map{WeylGroup, WeylGroup}[Map: W -> Weyl group, Map: W -> Weyl group])
+
+julia> emb[1](U[1][1]) == W[1]
+true
+
+julia> proj[2](W[4]) == U[2][1]
+true
+```
+"""
+function irreducible_factors(W::WeylGroup; morphisms::Bool=false)
+  types, ordering = root_system_type_with_ordering(root_system(W))
+  num_factors = length(types) # Number of irreducible factors of W
+
+  U = Vector{WeylGroup}(undef, num_factors)
+  emb = Vector{Map{WeylGroup,WeylGroup}}(undef, num_factors)
+  proj = Vector{Map{WeylGroup,WeylGroup}}(undef, num_factors)
+
+  start_index = 1
+  for (i, type) in enumerate(types)
+    rk = type[2] # rank of the i-th irreducible factor
+    end_index = start_index + rk - 1
+    U[i], emb[i], proj[i] = parabolic_subgroup_with_projection(
+      W, ordering[start_index:end_index]; check=false # No check needed by construction
+    )
+    start_index = end_index + 1
+  end
+
+  if morphisms
+    return U, emb, proj
+  else
+    return U
+  end
+end
+
+@doc raw"""
+    inner_direct_product(L::AbstractVector{WeylGroup}; morphisms::Bool=false)
+    inner_direct_product(L::WeylGroup...; morphisms::Bool=false)
+
+If `morphisms` is `false`, then return a Weyl group `W` that is isomorphic to the
+direct product of the Weyl groups in `L`.
+The generators of `W` are in natural 1-1 correspondence with the generators of the groups in `L`
+in the given order.
+
+If `morphisms` is `true`, return a triple (`W`, `emb`, `proj`) where
+`W` is as above and `emb`, `proj` are vectors such that
+`emb[i]` (resp., `proj[i]`) is the embedding of `L[i]` into `W`
+(resp., the projection of `W` onto `L[i]`).
+
+See also [`inner_direct_product(::AbstractVector{T})`](@ref).
+
+# Examples
+```jldoctest
+julia> W1 = weyl_group(:A, 2); W2 = weyl_group(:B, 3);
+
+julia> W, emb, proj = inner_direct_product(W1, W2; morphisms=true)
+(Weyl group of root system of type A2 x B3, Map{WeylGroup, WeylGroup}[Map: W1 -> W, Map: W2 -> W], Map{WeylGroup, WeylGroup}[Map: W -> W1, Map: W -> W2])
+
+julia> proj[2](W[3]) == W2[1]
+true
+
+julia> gens(W) == vcat(emb[1].(gens(W1)), emb[2].(gens(W2)))
+true
+```
+"""
+function inner_direct_product(L::AbstractVector{WeylGroup}; morphisms::Bool=false)
+  @req length(L) > 0 "The collection of groups must be non-empty"
+
+  cm_blocks = [cartan_matrix(factor) for factor in L]
+  cm = block_diagonal_matrix(cm_blocks)
+  product = weyl_group(cm)
+
+  if !morphisms
+    return product
+  end
+
+  emb = Vector{Map{WeylGroup,WeylGroup}}(undef, length(L))
+  proj = Vector{Map{WeylGroup,WeylGroup}}(undef, length(L))
+  start_index = 1
+  for i in eachindex(L)
+    factor = L[i]
```

Comment:
```suggestion
  for (i, factor) in enumerate(L)
```
a bit more julia-style